include/sound/soc.h

```diff
@@ -1159,6 +1159,23 @@ struct snd_soc_pcm_runtime {
 #define for_each_rtd_codec_dai_rollback(rtd, i, dai)		\
 	for (; ((--i) >= 0) && ((dai) = rtd->codec_dais[i]);)
 
+/**
+ * snd_soc_fw_mach: common fields for DT/ACPI machine descriptor.
+ *
+ * @drv_name: machine driver name
+ * @fw_filename: firmware file name. Used when SOF is not enabled.
+ * @sof_fw_filename: Sound Open Firmware file name, if enabled
+ * @sof_tplg_filename: Sound Open Firmware topology file name, if enabled
+ * @pdata: intended for platform data or machine specific-ops. This structure
+ *  is not constant since this field may be updated at run-time
+ */
+struct snd_soc_fw_mach {
+	const char *drv_name;
+	const char *fw_filename;
+	const char *sof_fw_filename;
+	const char *sof_tplg_filename;
+	void *pdata;
+};
```

Comment:
@dbaluta It is my understanding that in the DT/OF world, you have a 'handle' for the machine driver, so the drv_name is not useful since you will have a different mechanism to probe that machine driver. You could also make all these fields defined in that DT blob.

I think we'd need to step back a little and check what steps you need to complete before reshuffling all the data structures. No one on the Intel side has a clear understanding of the DT/OF setup, so we should agree on that first, maybe with some ascii art showing the sequences.

Reply:
@dbaluta indeed there a handle for the machine driver. I like your idea.

Indeed there is this function

``of_platform_populate() - Populate platform_devices from device tree data``

so we can create the platform device using this instead of Intel's platform platform_device_register_data. We need to modify core.c maybe add some device ops for this.

Will try to come up with a drawing, so far I've tried to emulate as much as possible the existing implementation for Intel boards, but obviously there a lots of differences between DT / ACPI.